### PR TITLE
Implement setCookie and getCookieValue

### PR DESCRIPTION
**Pull request by senthalan**
## Purpose
- Implement setCookie and getCookieValue

### setCookie(response, name, value, properties)
Sets a cookie named ‘name’ with the value as ‘value’. ‘Properties’ is a map that may contain optional attributes of the cookie with two other custom attributes ‘encrypt’ and‘sign’.

If ‘sign’ is true, then the value will be signed. Default value will be false.
If ‘encrypt’ is true, then value will be encrypted. Default value will be false.

Here is a sample function
```
setCookie(context.response, "name", "test", {"max-age" : 4000,
						"path" : "/",
						"domain" : "localhost",
						"httpOnly" : true,
						"secure" : true,
						"version" : 1,
						"comment" : "some comments",
						"encrypt" : true,
						"sign" : true}) 
```

Size of the value has to be less than the RSA key pair length if the encrypt is enabled.

### getCookieValue(request, name, properties)
Get the plan text cookie value for cookie ‘name’ if present.  ‘Properties’ is an optional map that may contain optional attributes ‘decrypt’ and ‘validateSignature’.

If ‘validateSignature’ is true, then the signature will be validated before returning. Default value will be false.
If ‘decrypt’ is true, then value will be decrypted. Default value will be false.

`getCookieValue(context.request,"name", {"decrypt" : true,"validateSignature" : true })
`

### CODE REVIEW

**Comment by pulasthi7** [comment]
@senthalan As this is a new component, you need to add this to the product also. Refer https://github.com/wso2/product-is/pull/3181 and create a separate PR for that also.

**Comment by senthalan** [comment]
PR to add to the product : https://github.com/wso2/product-is/pull/3187

File: components/org.wso2.carbon.identity.conditional.auth.functions.cookie/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/cookie/CookieFunctionImpl.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            try {
+                signature = Base64.encode(SignatureUtil.doSignature(value));
+            } catch (Exception e) {
+                log.error("Error occurred when signing the cookie value.", e);
+                return;
+            }
+        }
+        if (encrypt) {
+            try {
+                value = CryptoUtil.getDefaultCryptoUtil().encryptAndBase64Encode(Base64.decode(value));
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.conditional.auth.functions.cookie/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/cookie/CookieFunctionImpl.java]: why do we need to base 64 decode the value? value is not b64 encoded at this time, right?

Comment (pulasthi7) [inline components/org.wso2.carbon.identity.conditional.auth.functions.cookie/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/cookie/CookieFunctionImpl.java]: Also, shall we check for the value length before trying to encrypt

Comment (senthalan) [reply components/org.wso2.carbon.identity.conditional.auth.functions.cookie/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/cookie/CookieFunctionImpl.java]: encryptAndBase64Encode() method expects bytes[] array as input.  And base64DecodeAndDecrypt() methods Base64.decode() before decrypting. 

Due to this reason, we need to use Base64.decode() other than .getBytes(). Else decryption gives different value.

Comment (senthalan) [reply components/org.wso2.carbon.identity.conditional.auth.functions.cookie/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/cookie/CookieFunctionImpl.java]: maximum allowed value length will verify with the key length. If the value length is larger than the key length, the encryptAndBase64Encode() will throw an exception. We caught it and log the error.

**Pull request merged**